File: R/websocket.R
#' @useDynLib websocket, .registration = TRUE
#' @import later
#' @importFrom Rcpp sourceCpp
#' @importFrom R6 R6Class
#' @include RcppExports.R
NULL

# Used to "null out" handler functions after a websocket client is closed
null_func <- function(...) { }

#' Create a WebSocket client
#'
#' \preformatted{
#' WebSocket$new(url,
#'   protocols = character(0),
#'   headers = NULL,
#'   autoConnect = TRUE,
#'   accessLogChannels = c("none"),
#'   errorLogChannels = NULL,
#'   maxMessageSize = 32 * 1024 * 1024)
#' }
#'
#' @details
#'
#' A WebSocket object has four events you can listen for, by calling the
#' corresponding `onXXX` method and passing it a callback function. All callback
#' functions must take a single `event` argument. The `event` argument is a
#' named list that always contains a `target` element that is the WebSocket
#' object that originated the event, plus any other relevant data as detailed
#' below.
#'
#' \describe{
#'   \item{\code{onMessage}}{Called each time a message is received from the
#'     server. The event will have a `data` element, which is the message
#'     content. If the message is text, the `data` will be a one-element
#'     character vector; if the message is binary, it will be a raw vector.}
#'   \item{\code{onOpen}}{Called when the connection is established.}
#'   \item{\code{onClose}}{Called when a previously-opened connection is closed.
#'     The event will have `code` (integer) and `reason` (one-element character)
#'     elements that describe the remote's reason for closing.}
#'   \item{\code{onError}}{Called when the connection fails to be established.
#'     The event will have an `message` element, a character vector of length 1
#'     describing the reason for the error.}
#' }
#'
#' Each `onXXX` method can be called multiple times to register multiple
#' callbacks. Each time an `onXXX` is called, its (invisible) return value is a
#' function that can be invoked to cancel that particular registration.
#'
#' A WebSocket object also has the following methods:
#'
#' \describe{
#'   \item{\code{connect()}}{Initiates the connection to the server. (This does
#'     not need to be called unless you have passed `autoConnect=FALSE` to the
#'     constructor.)}
#'   \item{\code{send(msg)}}{Sends a message to the server.}
#'   \item{\code{close()}}{Closes the connection.}
#'   \item{\code{readyState()}}{Returns an integer representing the state of the
#'     connection.
#'     \describe{
#'       \item{\code{0L}: Connecting}{The WebSocket has not yet established a
#'       connection with the server.}
#'       \item{\code{1L}: Open}{The WebSocket has connected and can send and
#'       receive messages.}
#'       \item{\code{2L}: Closing}{The WebSocket is in the process of closing.}
#'       \item{\code{3L}: Closed}{The WebSocket has closed, or failed to open.}
#'     }}
#'   \item{setAccessLogChannels(channels)}{Enable the websocket Access channels after the
#'     websocket's creation.  A value of \code{NULL} will not enable any new Access channels.}
#'   \item{setErrorLogChannels(channels)}{Enable the websocket Error channels after the
#'     websocket's creation.  A value of \code{NULL} will not enable any new Error channels.}
#'   \item{clearAccessLogChannels(channels)}{Disable the websocket Access channels after the
#'     websocket's creation.  A value of \code{NULL} will not clear any existing Access channels.}
#'   \item{clearErrorLogChannels(channels)}{Disable the websocket Error channels after the
#'     websocket's creation.  A value of \code{NULL} will not clear any existing Error channels.}
#' }
#'
#' @param url The WebSocket URL. Should begin with \code{ws://} or \code{wss://}.
#' @param protocols Zero or more WebSocket sub-protocol names to offer to the server
#'   during the opening handshake.
#' @param headers A named list or character vector representing keys and values
#'   of headers in the initial HTTP request.
#' @param autoConnect If set to `FALSE`, then constructing the WebSocket object
#'   will not automatically cause the connection to be established. This can be
#'   used if control will return to R before event handlers can be set on the
#'   WebSocket object (i.e. you are constructing a WebSocket object manually at
#'   an interactive R console); after you are done attaching event handlers, you
#'   must call `ws$connect()` to establish the WebSocket connection.
#' @param accessLogChannels A character vector of access log channels that are
#'   enabled.  Defaults to \code{"none"}, which displays no normal, websocketpp logging activity.
#'   Setting \code{accessLogChannels = NULL} will use default websocketpp behavior.
#'   Multiple access logging levels may be passed in for them to be enabled.
#'
#'   A few commonly used access logging values are:
#'   \describe{
#'     \item{\code{"all"}}{Special aggregate value representing "all levels"}
#'     \item{\code{"none"}}{Special aggregate value representing "no levels"}
#'     \item{\code{"rerror"}}{Recoverable error. Recovery may mean cleanly closing the connection
#'           with an appropriate error code to the remote endpoint.}
#'     \item{\code{"fatal"}}{Unrecoverable error. This error will trigger immediate unclean
#'           termination of the connection or endpoint.}
#'   }
#'
#'   All logging levels are explained in more detail at \url{https://docs.websocketpp.org/reference_8logging.html}.
#' @param errorLogChannels A character vector of error log channels that are
#'   displayed.  The default value is \code{NULL}, which will use default websocketpp behavior.
#'   Multiple error logging levels may be passed in for them to be enabled.
#'
#'   A few commonly used error logging values are:
#'   \describe{
#'     \item{\code{"all"}}{Special aggregate value representing "all levels"}
#'     \item{\code{"none"}}{Special aggregate value representing "no levels"}
#'     \item{\code{"connect"}}{One line for each new connection that is opened}
#'     \item{\code{"disconnect"}}{One line for each new connection that is closed}
#'   }
#'
#'   All logging levels are explained in more detail at \url{https://docs.websocketpp.org/reference_8logging.html}.
#' @param maxMessageSize The maximum size of a message in bytes. If a message
#'   larger than this is sent, the connection will fail with the \code{message_too_big}
#'   protocol error.
#'
#'
#' @name WebSocket
#'
#' @examples
#' ## Only run this example in interactive R sessions
#' if (interactive()) {
#'
#' # Create a websocket using the websocket.org test server
#' ws <- WebSocket$new("ws://echo.websocket.org/")
#' ws$onMessage(function(event) {
#'   cat("Client got msg:", event$data, "\n")
#' })
#' ws$onClose(function(event) {
#'   cat("Client disconnected\n")
#' })
#' ws$onOpen(function(event) {
#'   cat("Client connected\n")
#' })
#'
#' # Try sending a message with ws$send("hello").
#' # Close the websocket with ws$close() after you're done with it.
#' }
NULL

#' @export
WebSocket <- R6Class("WebSocket",
  public = list(
    initialize = function(url,
      protocols = character(0),
      headers = NULL,
      autoConnect = TRUE,
      accessLogChannels = c("none"),
      errorLogChannels = NULL,
      maxMessageSize = 32 * 1024 * 1024,
      loop = later::current_loop()
    ) {
      private$callbacks <- new.env(parent = emptyenv())
      private$callbacks$open <- Callbacks$new()
      private$callbacks$close <- Callbacks$new()
      private$callbacks$error <- Callbacks$new()
      private$callbacks$message <- Callbacks$new()

      if (length(maxMessageSize) != 1 || !is.numeric(maxMessageSize) || maxMessageSize < 0){
        stop("maxMessageSize must be a non-negative integer")
      }

      private$wsObj <- wsCreate(
        url, loop$id, self, private,
        private$accessLogChannels(accessLogChannels, "none"),
        private$errorLogChannels(errorLogChannels, "none"),
        maxMessageSize
      )

      mapply(names(headers), headers, FUN = function(key, value) {
        wsAppendHeader(private$wsObj, key, value)
      })
      wsAddProtocols(private$wsObj, protocols)

      private$pendingConnect <- TRUE
      if (autoConnect) {
        self$connect()
      }
    },
    connect = function() {
      if (private$pendingConnect) {
        private$pendingConnect <- FALSE
        wsConnect(private$wsObj)
      } else {
        warning("Ignoring extraneous connect() call (did you mean to have autoConnect=FALSE in the constructor?)")
      }
    },
    readyState = function() {
      code <- function(value, desc) {
        structure(value, description = desc)
      }

      if (private$pendingConnect) {
        return(code(-1L, "Pre-connecting"))
      }

      switch(wsState(private$wsObj),
        INIT = code(0L, "Connecting"),
        OPEN = code(1L, "Open"),
        CLOSING = code(2L, "Closing"),
        CLOSED = code(3L, "Closed"),
        FAILED = code(3L, "Closed"),
        stop("Unknown state ", wsState(private$wsObj))
      )
    },
    onOpen = function(callback) {
      invisible(private$callbacks[["open"]]$register(callback))
    },
    onClose = function(callback) {
      invisible(private$callbacks[["close"]]$register(callback))
    },
    onError = function(callback) {
      invisible(private$callbacks[["error"]]$register(callback))
    },
    onMessage = function(callback) {
      invisible(private$callbacks[["message"]]$register(callback))
    },
    protocol = function() {
      wsProtocol(private$wsObj)
    },
    send = function(msg) {
      wsSend(private$wsObj, msg)
    },
    close = function(code = 1000L, reason = "") {
      wsClose(private$wsObj, code, reason)
    },
    setAccessLogChannels = function(channels = c("all")) {
      wsUpdateLogChannels(private$wsObj, "access", "set", private$accessLogChannels(channels, "none"))
    },
    setErrorLogChannels = function(channels = c("all")) {
      wsUpdateLogChannels(private$wsObj, "error", "set", private$errorLogChannels(channels, "none"))
    },
    clearAccessLogChannels = function(channels = c("all")) {
      wsUpdateLogChannels(private$wsObj, "access", "clear", private$accessLogChannels(channels, "all"))
    },
    clearErrorLogChannels = function(channels = c("all")) {
      wsUpdateLogChannels(private$wsObj, "error", "clear", private$errorLogChannels(channels, "all"))
    }
  ),
  private = list(
    wsObj = NULL,
    callbacks = NULL,
    pendingConnect = TRUE,
    getInvoker = function(eventName) {
      callbacks <- private$callbacks[[eventName]]
      stopifnot(!is.null(callbacks))
      callbacks$invoke
    },
    accessLogChannelValues = c(
      "none", "connect", "disconnect", "control", "frame_header", "frame_payload",
      "message_header", "message_payload", "endpoint", "debug_handshake", "debug_close", "devel",
      "app", "http", "fail", "access_core", "all"
    ),
    errorLogChannelValues = c("none", "devel", "library", "info", "warn", "rerror", "fatal", "all"),
    accessLogChannels = function(channels, stompValue) {
      if (is.null(channels)) return(character(0))
      channels <- match.arg(channels, private$accessLogChannelValues, several.ok = TRUE)
      if (stompValue %in% channels) channels <- stompValue
      channels
    },
    errorLogChannels = function(channels, stompValue) {
      if (is.null(channels)) return(character(0))
      channels <- match.arg(channels, private$errorLogChannelValues, several.ok = TRUE)
      if (stompValue %in% channels) channels <- stompValue
      channels
    }
  )
)

Callbacks <- R6Class(
  'Callbacks',
  private = list(
    .nextId = integer(0),
    .callbacks = 'environment'
  ),
  public = list(
    initialize = function() {
      # NOTE: we avoid using '.Machine$integer.max' directly
      # as R 3.3.0's 'radixsort' could segfault when sorting
      # an integer vector containing this value
      private$.nextId <- as.integer(.Machine$integer.max - 1L)
      private$.callbacks <- new.env(parent = emptyenv())
    },
    register = function(callback) {
      if (!is.function(callback)) {
        stop("callback must be a function")
      }
      id <- as.character(private$.nextId)
      private$.nextId <- private$.nextId - 1L
      private$.callbacks[[id]] <- callback
      return(function() {
        rm(list = id, pos = private$.callbacks)
      })
    },
    invoke = function(...) {
      # Ensure that calls are invoked in the order that they were registered
      keys <- as.character(sort(as.integer(ls(private$.callbacks)), decreasing = TRUE))
      callbacks <- mget(keys, private$.callbacks)

      for (callback in callbacks) {
        tryCatch(
          callback(...),
          error = function(e) {
            message("Error in websocket callback: ", e$message)
          },
          interrupt = function(e) {
            message("Interrupt received while executing websocket callback.")
          }
        )

      }
    },
    count = function() {
      length(ls(private$.callbacks))
    }
  )
)

File: src/websocket.cpp
#ifdef _WIN32
#include <winsock2.h>
#include <windows.h>

// These are defined by windows.h but we have to undefine them so that the
// typedef enum Rboolean will be used later on.
#undef TRUE
#undef FALSE

#endif // _WIN32


#include <iostream>
#include "cpp11.hpp"
#include "wrapped_print.h"
#include <websocketpp/common/functional.hpp>
#include "client.hpp"
#include "websocket_defs.h"
#include "websocket_task.h"
#include "websocket_connection.h"
#include "debug.h"


shared_ptr<WebsocketConnection> xptrGetWsConn(SEXP wsc_xptr) {
  if (TYPEOF(wsc_xptr) != EXTPTRSXP) {
    cpp11::stop("Expected external pointer.");
  }
  return *reinterpret_cast<shared_ptr<WebsocketConnection>*>(R_ExternalPtrAddr(wsc_xptr));
}

void wsc_deleter(SEXP wsc_xptr) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  if (!wsc->client->stopped()) {
    // I don't think we'll ever actually get here because as long as the
    // connection is open, the WebsocketConnection object will contain a
    // reference to the R6 object (robjPublic), which contains a reference
    // back to the WebsocketConnection object. This prevents the R6 object
    // from getting GC'd, and its external pointer in turn will not get GC'd.
    // But just in case we get here somehow, we'll stop the thread.
    wsc->client->stop();
  }
  delete reinterpret_cast<shared_ptr<WebsocketConnection>*>(R_ExternalPtrAddr(wsc_xptr));
  R_ClearExternalPtr(wsc_xptr);
}


[[cpp11::register]]
SEXP wsCreate(
  std::string uri,
  int loop_id,
  cpp11::environment robjPublic,
  cpp11::environment robjPrivate,
  cpp11::strings accessLogChannels,
  cpp11::strings errorLogChannels,
  int maxMessageSize
) {
  REGISTER_MAIN_THREAD()
  WebsocketConnection* wsc = new WebsocketConnection(
    uri, loop_id, robjPublic, robjPrivate, accessLogChannels, errorLogChannels, maxMessageSize
  );

  shared_ptr<WebsocketConnection> *wsc_pp = new shared_ptr<WebsocketConnection>(wsc);
  SEXP wsc_xptr = PROTECT(R_MakeExternalPtr(wsc_pp, R_NilValue, R_NilValue));
  R_RegisterCFinalizerEx(wsc_xptr, wsc_deleter, TRUE);
  UNPROTECT(1);
  return wsc_xptr;
}

[[cpp11::register]]
void wsAppendHeader(SEXP wsc_xptr, std::string key, std::string value) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  wsc->client->append_header(key, value);
}

[[cpp11::register]]
void wsAddProtocols(SEXP wsc_xptr, cpp11::strings protocols) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  for (auto it = protocols.begin(); it != protocols.end(); ++it) {
    std::string protocol = *it;
    wsc->client->add_subprotocol(protocol);
  }
}

[[cpp11::register]]
void wsAddProxy(SEXP wsc_xptr, std::string proxy_url) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  wsc->client->set_proxy(proxy_url);
}

[[cpp11::register]]
void wsConnect(SEXP wsc_xptr) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);

  wsc->client->connect();

  // Starts a new thread in which WebsocketTask::execute is called. Object is
  // automatically deleted when thread stops.
  WebsocketTask* wst = new WebsocketTask(wsc);
  wst->begin();
}

[[cpp11::register]]
void wsSend(SEXP wsc_xptr, SEXP msg) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);

  if (TYPEOF(msg) == STRSXP &&
      Rf_length(msg) == 1 &&
      STRING_ELT(msg, 0) != NA_STRING)
  {
    // TODO: Make sure that message lifetime is long enough
    const char* msg_ptr = CHAR(STRING_ELT(msg, 0));
    int len = R_nchar(STRING_ELT(msg, 0), Bytes, FALSE, FALSE, "wsSend");
    // When send() is called, I believe that the message is immediately
    // appended to the message buffer, and then the actual sending of the
    // message is queued. Since a copy of the message is made, it should be
    // safe if the original msg is cleared before the actual send happens.
    wsc->client->send(msg_ptr, len, ws_websocketpp::frame::opcode::text);

  } else if (TYPEOF(msg) == RAWSXP) {
    wsc->client->send(RAW(msg), Rf_length(msg), ws_websocketpp::frame::opcode::binary);
  } else {
    cpp11::stop("msg must be a one-element character vector or a raw vector.");
  }
}

[[cpp11::register]]
void wsClose(SEXP wsc_xptr, uint16_t code, std::string reason) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  wsc->close(code, reason);
}

[[cpp11::register]]
std::string wsProtocol(SEXP wsc_xptr) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  return wsc->client->get_subprotocol();
}

[[cpp11::register]]
std::string wsState(SEXP wsc_xptr) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  switch(wsc->state) {
    case WebsocketConnection::STATE::INIT: return "INIT";
    case WebsocketConnection::STATE::OPEN: return "OPEN";
    case WebsocketConnection::STATE::CLOSING: return "CLOSING";
    case WebsocketConnection::STATE::CLOSED: return "CLOSED";
    case WebsocketConnection::STATE::FAILED: return "FAILED";
  }

  // Shouldn't be possible to get here, but some compilers still complain
  // about reaching end of a non-void function.
  return "UNKNOWN";
}

[[cpp11::register]]
void wsUpdateLogChannels(
  SEXP wsc_xptr,
  std::string accessOrError,
  std::string setOrClear,
  cpp11::strings logChannels
) {
  ASSERT_MAIN_THREAD()
  shared_ptr<WebsocketConnection> wsc = xptrGetWsConn(wsc_xptr);
  wsc->client->update_log_channels(accessOrError, setOrClear, logChannels);
}
